driver-core/src/main/com/mongodb/internal/logging/StructuredLogger.java

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.internal.logging;
+
+import com.mongodb.connection.ClusterId;
+import com.mongodb.internal.VisibleForTesting;
+import com.mongodb.internal.diagnostics.logging.Logger;
+import com.mongodb.internal.diagnostics.logging.Loggers;
+import com.mongodb.internal.logging.StructuredLogMessage.Entry;
+import com.mongodb.internal.logging.StructuredLogMessage.Level;
+import com.mongodb.lang.Nullable;
+
+import java.util.concurrent.ConcurrentHashMap;
+
+import static com.mongodb.internal.VisibleForTesting.AccessModifier.PRIVATE;
+import static java.lang.String.format;
+
+/**
+ * <p>This class is not part of the public API and may be removed or changed at any time</p>
+ */
+public final class StructuredLogger {
+
+    private static final ConcurrentHashMap<String, StructuredLoggingInterceptor> INTERCEPTORS = new ConcurrentHashMap<>();
+
+    private final Logger logger;
+
+    @VisibleForTesting(otherwise = PRIVATE)
+    public static void addInterceptor(final String clusterDescription, final StructuredLoggingInterceptor interceptor) {
+        INTERCEPTORS.put(clusterDescription, interceptor);
+    }
+
+    @VisibleForTesting(otherwise = PRIVATE)
+    public static void removeInterceptor(final String clusterDescription) {
+        INTERCEPTORS.remove(clusterDescription);
+    }
+
+    @Nullable
+    private static StructuredLoggingInterceptor getInterceptor(@Nullable final String clusterDescription) {
+        if (clusterDescription == null) {
+            return null;
+        }
+        return INTERCEPTORS.get(clusterDescription);
+    }
+
+    public StructuredLogger(final String suffix) {
+        this(Loggers.getLogger(suffix));
+    }
+
+    @VisibleForTesting(otherwise = PRIVATE)
+    public StructuredLogger(final Logger logger) {
+        this.logger = logger;
+    }
+
+    public boolean isRequired(final Level level, final ClusterId clusterId) {
+        if (getInterceptor(clusterId.getDescription()) != null) {
+            return true;
+        }
+
+        //noinspection SwitchStatementWithTooFewBranches
+        switch (level) {
+            case DEBUG:
+                return logger.isDebugEnabled();
+            default:
+                throw new UnsupportedOperationException();
+        }
+   }
+
+    public void log(final StructuredLogMessage message, final String format) {
```

Comment:
This is looking cleaner ;)

I think we can we negate the cost of always creating a StructuredLogMessage - by just providing a supplier. That way its only created in the testing environments where there is an interceptor.

Would require decoupling the logging message from the StructuredLogMessage.

Something like:

```
public void log(final Consumer<Logger> loggingConsumer, final Supplier<StructuredLogMessage> structuredLogMessageSupplier)
```

This way the cost only has to be applied if either logger.isDebugEnabled and/or there is an inteceptor.

Reply:
Given that almost everywhere I look in the code we already guard with `is<Level>Enabled`, I'm inclined to just leave it like this.  But I'm not sure what the intent of the `loggingConsumer` is in your example.  Can you explain how that would be used?

Reply:
My thoughts were instead of `LoggingCommandEventSender`having to be aware of the `StructuredLogger` requirements with the `loggingRequired` checks you could move all that inside.

Eg: 

```
// Change the method name to include the level
public void debug(final ClusterId clusterId, 
                  final Consumer<Logger> loggingConsumer, 
                  final Supplier<StructuredLogMessage> structuredLogMessageSupplier)
```

The usage becomes:

```
logger.debug(getClusterId(), 
             l -> { 
               l.debug(format("Sending command '%s' with request id %d to database %s on connection [%s] to server %s", 
               commandString, message.getId(), message.getNamespace().getDatabaseName(), 
               description.getConnectionId(), description.getServerAddress()));},
            () -> {
                // Create Entries
                List<Entry> ....
                // return the Structured Log Message
                return new StructuredLogMessage(COMMAND, DEBUG, "Command started", getClusterId(), entries);
            }  
);
```

That way all the guards can occur in the StructuredLogger and the StructuredLogMessage is only computed if its actually required.

Reply:
My first attempt at this did actually include a method per level but I didn't like how it required repeating the logging level in the `StructureLogMessage` parameter list, and then having to check for that to make sure they actually match.  That seemed odd to me so I changed it to a single `log` method

The other tricky thing is that sometimes (as in command logging) there is an expensive log entry value to compute, and if you have to do it twice, once in each lambda, there is no way to share the value between the two.  So we'd end up needing a conditional anyway to check whether logging is required, in order to create any expensive value.  Also, in general I think having to repeat the entry values is going to be cumbersome, especially in command logging where there are entries that are conditionally included based on whether they are null.

Another possibility to consider here is to emulate the [SLF4J 2.0 API](https://www.slf4j.org/api/org/slf4j/Logger.html), particularly `LoggingEventBuilder`.  Obviously that has been vetted pretty carefully.

Reply:
`LoggingEventBuilder` looks good. Perhaps in 5.0 we could remove our own logging wrapper and just use SLF4J?

Reply:
Definitely there are costs with either approach and both are cumbersome - lets move on and readdress in the future if required.